### DISCUSSION TITLE
lualine info if mason packages updates are available

### DISCUSSION THREAD

**Discussion by kekscode** [mason-org/mason.nvim#1535] [Q&A]
I use lualine and the lazy package manager. This combination has this little thingie:

<img width="90" alt="image" src="https://github.com/williamboman/mason.nvim/assets/803252/9b8cec85-9dac-4f15-800a-642853afb4c0">

which signals that new updates for lazy plugins are available.

I would like to have that for Mason. I know from the docs that mason has:

```lua
local registry = require("mason-registry")

registry.refresh(function ()
    local packages = registry.get_all_packages()
    -- print(vim.inspect(packages))
end)
```

And i feel that this could be used for the same purpose. Did anybody do the same thing or even got a working example?

**Comment by williamboman** [comment]
You probably want something like this

```lua
local registry = require("mason-registry")

local function update_lualine()
    local installed_packages = registry.get_installed_package_names()
    local count = #installed_packages
   -- do something with count
end

registry.refresh(update_lualine)

registry:on("package:install:success", update_lualine)
registry:on("package:uninstall:success", update_lualine)
```

**Reply by kekscode** [reply]
Thank  you for the code example, appreciated. It is nice to get the count of all installed packages, but i wonder how to get the count of outdated packages like in https://github.com/williamboman/mason.nvim/blob/cd7835b15f5a4204fc37e0aa739347472121a54c/lua/mason/ui/components/main/package_list.lua#L163-L165 eventually  showing the count of outdated  mason packages  in  lualine alongside the count of outdated  packages from  the lazy package manager.

Ii found the `check_new_version` callback and i think i can figure out how to loop over things (i don't know much about lua beyond configuring plugin setup calls...), but i couldn't stitch it together just yet. I'll post a solution when i figure it out.

**Reply by kekscode** [reply]
ok, i think i have it working now:

<img width="90" alt="image" src="https://github.com/williamboman/mason.nvim/assets/803252/c2f75949-4a35-4322-b159-980f88adf037">

7 updates detected. If i click on it, it does `:Mason`.

The check function is:

```lua
-- check for mason package upgrades
local function lualine_mason_updates()
  local registry = require("mason-registry")
  local installed_packages = registry.get_installed_package_names()
  local upgrades_available = false
  local packages_outdated = 0
  function myCallback(success, result_or_err)
    if success then
      upgrades_available = true
      packages_outdated = packages_outdated + 1
    end
  end

  for _, pkg in pairs(installed_packages) do
    local p = registry.get_package(pkg)
    if p then
      p:check_new_version(myCallback)
    end
  end

  if upgrades_available then
    return packages_outdated
  else
    return ""
  end
end
```

And this is how i embed it in my lualine:

```lua
-- [...]
            {
              lualine_mason_updates,
              icon     = "🧱",
              on_click = function()
                vim.cmd("Mason")
              end,
            },
-- [...]
```

**Reply by Blackjacx** [reply]
WOW, cool I searched soooo long for a solution to this 🤩 Thank you so much! As you seem to know what you're doing here, can you also tell maybe why there is a [Mason extension](https://github.com/nvim-lualine/lualine.nvim/blob/master/lua/lualine/extensions/mason.lua) for Lualine and how it is supposed to work?

**Reply by kekscode** [reply]
Thanks @Blackjacx for the heads up, happy that the snippet is useful for you. As for the mason plugin for lualine, it is apparently ~10mo old, so i wasn’t aware of it’s existence back then. I‘ll have a look at it.

**Reply by kekscode** [reply]
@Blackjacx i had a look at the plugin. It triggers when `:Mason` is opened (recognized the filetype https://github.com/nvim-lualine/lualine.nvim/blob/master/lua/lualine/extensions/mason.lua#L27) and then shows you how many packages are installed by modifying the lualine accordingly:

<img width="247" alt="image" src="https://github.com/user-attachments/assets/8dcad6ec-bbfb-43de-8114-454fe2166f8c">

**Reply by Blackjacx** [reply]
Ahh cool 👍 Thx for checking that out 🙏

**Reply by afpfernandes** [reply]
For those interested, `check_new_version()` was deprecated and the same can be achieved with `get_installed_version()` and `get_latest_version()`:

```lua
local function lualine_mason_updates()
  local registry = require 'mason-registry'
  local installed_packages = registry.get_installed_package_names()
  local packages_outdated = 0

  for _, pkg_name in ipairs(installed_packages) do
    local pkg = registry.get_package(pkg_name)
    if pkg then
      local ok_installed, installed_version = pcall(pkg.get_installed_version, pkg)
      local ok_latest, latest_version = pcall(pkg.get_latest_version, pkg)
      if ok_installed and ok_latest and installed_version ~= latest_version then
        packages_outdated = packages_outdated + 1
      end
    end
  end

  return packages_outdated > 0 and tostring(packages_outdated) or ''
end
```
